### ISSUE TITLE
Create `validation_time_split()`, similar to `validation_split()` but for first `prop` samples

### ISSUE THREAD

**Issue by simonschoe** [tidymodels/rsample#251]
Hi there,

I am facing the following setting and wondering if there is a more elegant/appropriate `rsample`-based solution to it. Suppose I have time-series data where each observation has a unique date (`date`) and a unique year end (`year_end`):
```r
df <- 
  tibble(
    date = c("2015-05-15", "2015-02-15", "2015-09-15", "2016-01-15", "2016-07-15", "2017-08-15", "2018-01-15", "2018-01-15", "2018-01-15", "2018-07-15", "2018-07-15"),
    year_end = c("2015-12-31", "2015-12-31", "2015-12-31", "2016-12-31", "2016-12-31", "2017-12-31", "2018-12-31", "2018-12-31", "2018-12-31", "2018-12-31", "2018-12-31"),
    features = rep(123, 11)
  ) %>% 
  mutate(across(c(date, year_end), as.Date))

df
```
```r
# A tibble: 11 x 3
   date       year_end   features
   <date>     <date>        <dbl>
 1 2015-05-15 2015-12-31      123
 2 2015-02-15 2015-12-31      123
 3 2015-09-15 2015-12-31      123
 4 2016-01-15 2016-12-31      123
 5 2016-07-15 2016-12-31      123
 6 2017-08-15 2017-12-31      123
 7 2018-01-15 2018-12-31      123
 8 2018-01-15 2018-12-31      123
 9 2018-01-15 2018-12-31      123
10 2018-01-15 2018-12-31      123
11 2018-01-15 2018-12-31      123
```
I'd now like to generate time-based resamples in a rolling window-like fashion, using two years as training and one year as test data:
```r
time_splits <- 
  sliding_period(
    df,
    year_end,
    "year",
    lookback = 1,
    assess_stop = 1,
    complete = T,
    skip = 0
  )
```
```r
# Sliding period resampling 
# A tibble: 2 x 2
  splits        id    
  <list>        <chr> 
1 <split [5/1]> Slice1
2 <split [3/5]> Slice2
```
Now, since I'd like to train more complex models on the data, I need to sub-divide the test portion of the split into a validation and test set. Ideally, I would want the data points relating to the first half year in the validation set and the data points relating to the second half year in the test set. Functions like `validation_split` did not really help here, since I don't want a random split but basically a split based on `date` > `20xx-06-30`. My current solution looks like this:
 ```r
val_test_split <- function(rset) {
  
  data <- rset$data
  
  val_ids <- 
    data %>% 
    transmute(index = ((row_number() %in% rset$out_id) & (lubridate::month(date) < 7))) %>% 
    pull %>% 
    which(arr.ind = T)
  
  test_ids <- 
    rset$out_id[!(rset$out_id %in% val_ids)]
    
  ind <- list(
    analysis = rset$in_id,
    assessment = val_ids
  )
  
  list(
    train_val_split = make_splits(ind, data),
    test_set = data[test_ids,]
  )
  
}

time_splits %>% 
  mutate(
    new_split = map(splits, ~val_test_split(.x)$train_val_split),
    test_set = map(splits, ~val_test_split(.x)$test_set)
  )
```
```r
# Sliding period resampling 
# A tibble: 2 x 4
  splits        id     new_split     test_set        
  <list>        <chr>  <list>        <list>          
1 <split [5/1]> Slice1 <split [5/0]> <tibble [1 x 3]>
2 <split [3/5]> Slice2 <split [3/3]> <tibble [2 x 3]>
```
However, I am obviously not super happy about this workaround. Is there a more elegant solution to this? I am thinking this must be a quite regular use case. Otherwise, we can maybe make a feature request out of this? :)

**Comment by juliasilge** [comment]
Thanks for this feedback @simonschoe!

The way you have described this doesn't quite sound like it aligns with how we design for data resampling strategies. I imagine you've already [looked at this section of our book](https://www.tmwr.org/resampling.html#resampling-methods), where we outline different resampling methods, including [how we treat a validation set](https://www.tmwr.org/resampling.html#validation).

![image](https://user-images.githubusercontent.com/12505835/129637449-12bf6670-2829-4c40-a518-4e88edf802a0.png)

We also have support for [nested resampling](https://www.tidymodels.org/learn/work/nested-resampling/), although that sounds like it is the opposite of what you are trying to do (we resample each analysis/training set again, to separate tuning from estimating performance).

Can you describe in a bit more detail what it is you are trying to do, and why you are taking this approach?

**Comment by simonschoe** [comment]
Hi @juliasilge, thanks for the reply and the included references. Of course, I've already digested the related materials from your book (great work and resource by the way! 🤩 ). Let me try to outline what I intend to do a little more:

In general, I have data with a strong time-series component on which I like to fit models. Due to the time-series component (especially due to lagged/rolling window based features), I can't employ random resampling schemes (like k-fold CV) as they would introduce data leakage as far as I am concerned (a validation or test set observation that lies before the training set observations on the time line doesn't make sense). Accordingly, I picked `sliding_period()`-based resampling envisioning the following setup:

![grafik](https://user-images.githubusercontent.com/53626067/129677499-2291fbed-95ab-4786-87e1-a9f91b65768e.png)

In my opinion this approach is not at odds with how you view the [validation set approach](https://www.tmwr.org/resampling.html#validation). Only difference is that I split the validation set from the test set year above instead of from the training set years. And, as already outlined, I need the splits to be time-based instead of random. As a result, I should be able to tune models over seven slices/resamples of data (assuming the dataset spans the 2010-2018 period), while my final out-of-sample performance estimate would be the average over the seven half-year test sets. Is this approach reasonable or do you see a conceptual error here? Isn't this also somewhat related to nested resampling where you tune over the inner loop and construct unbiased performance estimates over the outer loop (not sure if I got the essence of the nested resampling vignette tho).

**Comment by juliasilge** [comment]
It sounds like what you want to do is closest to nested resampling; check out [this page from the mlr folks](https://mlr.mlr-org.com/articles/tutorial/nested_resampling.html) as well as the [resources here](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C7&q=%22nested+resampling%22+inner+outer&btnG=).

We don't have full support for nested resampling with the time-based resampling functions, or a `validation_time_split()` yet, so for now you will need to do a bit of this manually:

``` r
library(tidyverse)
library(rsample)
#> Warning: package 'rsample' was built under R version 4.1.1
data("Chicago", package = "modeldata")

chicago_splits <- 
  sliding_period(
    Chicago,
    date,
    "year",
    lookback = 2,
    assess_stop = 1
  )

chicago_splits %>%
  mutate(inner_resamples = map(splits, ~ initial_time_split(analysis(.))))
#> # Sliding period resampling 
#> # A tibble: 13 × 3
#>    splits             id      inner_resamples  
#>    <list>             <chr>   <list>           
#>  1 <split [1074/366]> Slice01 <split [805/269]>
#>  2 <split [1096/365]> Slice02 <split [822/274]>
#>  3 <split [1096/365]> Slice03 <split [822/274]>
#>  4 <split [1096/365]> Slice04 <split [822/274]>
#>  5 <split [1095/366]> Slice05 <split [821/274]>
#>  6 <split [1096/365]> Slice06 <split [822/274]>
#>  7 <split [1096/365]> Slice07 <split [822/274]>
#>  8 <split [1096/365]> Slice08 <split [822/274]>
#>  9 <split [1095/366]> Slice09 <split [821/274]>
#> 10 <split [1096/365]> Slice10 <split [822/274]>
#> 11 <split [1096/365]> Slice11 <split [822/274]>
#> 12 <split [1096/365]> Slice12 <split [822/274]>
#> 13 <split [1095/241]> Slice13 <split [821/274]>
```

<sup>Created on 2021-08-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Be sure to keep in mind this image for how we conceptualize a validation set (just a single resample):

![image](https://user-images.githubusercontent.com/12505835/129762354-8872e2a6-1653-4c34-a52a-d1cce0fe4da8.png)

These validation sets of the Chicago data come from a single resample of each "analysis" (i.e. training) set from the sliding resamples. This particular resampling strategy make sliding resamples with 3 years of training data + 1 year of testing data, and then creates another split of each analysis/training set so that in the inner loop 3/4 of the data is used for fitting and 1/4 is used for estimate performance; you can change all those.

**Comment by juliasilge** [comment]
🎯  **Still an action item here:**

Let's consider making a `validation_time_split()` that is similar to `validation_split()` in the way that `initial_time_split()` is similar to `initial_split()`. Then folks could do something like this:

``` r
library(rsample)
#> Warning: package 'rsample' was built under R version 4.1.1
nested_cv(mtcars, outside = vfold_cv(v = 3), inside = validation_split(prop = 0.8))
#> # Nested resampling:
#> #  outer: 3-fold cross-validation
#> #  inner: Validation Set Split (0.8/0.2) 
#> # A tibble: 3 × 3
#>   splits          id    inner_resamples           
#>   <list>          <chr> <list>                    
#> 1 <split [21/11]> Fold1 <validation_split [1 × 2]>
#> 2 <split [21/11]> Fold2 <validation_split [1 × 2]>
#> 3 <split [22/10]> Fold3 <validation_split [1 × 2]>
```

<sup>Created on 2021-08-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by simonschoe** [comment]
Thank you so much for the illustration, taking the time and always responding! I conclude that in my special case, the inner resamples of the nested resampling approach are simply train-val-splits whereby the outer resamples come from the `sliding_period()` approach.

I guess your proposed solution will work for me. However, what still isn't possible, is to split on a specific date instead of simply based on a proportion of the data (I assume `initial_time_split()` requires data ordered by the relevant date column? This did not become 100% clear to me from the docs, at least not those of `initial_time_split()`).

____

In addition, I'd like to highlight a quirk a came across (surely, this is my own error, but I needed quite some time to debug - maybe such an error could be handled more idk "verbosely"?):

My actual dataset looked like this after applying `sliding_period()`:
```r
# Sliding period resampling 
# A tibble: 10 x 2
   splits              id     
   <list>              <chr>  
 1 <split [3789/1339]> Slice01
 2 <split [4253/1517]> Slice02
 3 <split [4396/1233]> Slice03
 4 <split [4089/1177]> Slice04
 5 <split [3927/1319]> Slice05
 6 <split [3729/1254]> Slice06
 7 <split [3750/1066]> Slice07
 8 <split [3639/1294]> Slice08
 9 <split [3614/103]>  Slice09
10 <split [2463/0]>    Slice10
```
When I tried to run
```r
time_res %>%
  mutate(inner_resamples = map(splits, ~initial_time_split(analysis(.), prop = 0.9)))
```
I ran into the following error: 
````
`Error: Problem with `mutate()` column `inner_resamples`. i `inner_resamples = map(splits, ~initial_time_split(assessment(.), prop = 0.5))`. x `in_id` must be a positive integer vector.`
````

Essentially, this was due to one corrupt date in my data (corresponding to the year 2099) which led to the creation of the `Slice10` I assume. It's just that I was wondering why the function would even allow the creation of a split with zero observations in the test set. I would have expected `Slice10` to be omitted in the first place...

**Comment by topepo** [comment]
A few things that would be really helpful is to know what model you are using and what you would do with each of these holdout sets. 

In general, we treat a validation set as a single resample (as shown above). 

For cases where some holdout data should be used for things like early stopping, that get's handled inside the modeling function itself (here's an example for [xgboost](https://parsnip.tidymodels.org/reference/xgb_train.html)). The reason that we do it this way is that the user of an additional holdout is a bit of a special case. It is commonly used with certain boosting operations and neural networks but not for many other models. 

TBH I don't think that we will change the system to do three-way splits but we might be able to help you more if we had more information.

**Comment by juliasilge** [comment]
Check out using the [purrr adverbs like `safely()` and `possibly()`](https://purrr.tidyverse.org/reference/safely.html) to handle [problems like you mention](https://www.infoworld.com/article/3601124/error-handling-in-r-with-purrrs-possibly.html) in your data.

**Comment by simonschoe** [comment]
Thanks to you both for your replies! 🤩 

In essence, I'd like to let different models compete in a horse race: elastic net, random forest, xgboost, svm and potentially a neural network. The validation set is meant for hyperparameter tuning and model selection while the test set is meant for an unbiased estimate of model performance out-of-sample. Due to the time-series component in my data I refrained from using k-fold CV and instead chose the approach outlined above. My understanding was that as soon as hyperparameter tuning come into a play, a three-way split is necessary to 1) search for the optimal parameter configuration for each model (based on the validation set performance) and to 2) provide a truly unbiased model estimate out-of-sample (using the unseen data from the test set).

**Comment by juliasilge** [comment]
We were chatting about this more and what we recommend is:

- Create an **initial split** into testing and training using either [`initial_time_split()`](https://rsample.tidymodels.org/reference/initial_split.html) if you are OK with taking the last `prop` as your testing data or [`make_splits()`](https://rsample.tidymodels.org/dev/reference/make_splits.html) if you need your testing data to be something more carefully crafted like exactly the last six months. You'll use the (single) testing set from this split to estimate out-of-sample performance for any models you have at the end of your modeling analysis.
- Take the training data from that initial split and create **resampling folds** using `sliding_period()`.
- Use tuning functions on those resamples to tune hyperparameters and choose the best model.
- Once you have a final model or two, use the single testing set to estimate performance on new data. 

We don't recommend having a separate test set for each resampling fold, as the resampling folds themselves have analysis/assessment sets. You might check out Chapters [5](https://www.tmwr.org/splitting.html) and [10](https://www.tmwr.org/resampling.html) of our book to think about "spending your data budget" and what the purpose of these different splits are.

**Comment by simonschoe** [comment]
@topepo @juliasilge thanks again for your thoughtful answer. I think I see the point know where my approach sparked some confusion. 💡 

In essence, my chosen approach stemmed from the fact that I'd also like to observe the model performance for different time slices to infere whether model performance increased or degraded over time, thats why I intuitively set aside multiple test/holdout sets instead of only one holdout set at the end. That being said, I start to realize that a more suitable approach would probably be to use a simple train-val-test split and investigate each time slice separately (i.e. fit models for each time slice) instead of employing resampling to find an "average model" which performs sufficiently well across all time slices... Surely, that would come at disadvantage of not having uncertainty estimates (due to the lack of resampling). I need to think a little more about the problem at hand, but again, thank you for your continuous help so far!

**Comment by juliasilge** [comment]
Closed in #256 

Folks can how do this:

``` r
library(rsample)
nested_cv(mtcars, outside = vfold_cv(v = 3), inside = validation_time_split(prop = 0.8))
#> # Nested resampling:
#> #  outer: 3-fold cross-validation
#> #  inner: Validation Set Split (0.8/0.2) 
#> # A tibble: 3 × 3
#>   splits          id    inner_resamples           
#>   <list>          <chr> <list>                    
#> 1 <split [21/11]> Fold1 <validation_split [1 × 2]>
#> 2 <split [21/11]> Fold2 <validation_split [1 × 2]>
#> 3 <split [22/10]> Fold3 <validation_split [1 × 2]>
```

<sup>Created on 2021-09-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>
